### ISSUE TITLE
Reference to undefined maven_version global variable causes failure when strict_variables is true

### ISSUE THREAD

**Issue by af6140** [maestrodev/puppet-maven#53]
with strict_variables = true.

Triggers following error:

Error: Evaluation Error: Unknown variable: '::maven_version'. at /etc/puppet/modules/maven/manifests/maven.pp:42:6 on node centos-7-x64

**Comment by MikeCaspar** [comment]
I am having the same problem on a Ubuntu 14.04 machine. ( I suspect it is fairly universal ).

As a workaround, you could set the desired Maven Version in site.pp

* Caution: This will likely effect maven installations in your entire environment! *

in site.pp....
`
$maven_version = '3.2.5'
`

**Comment by MikeCaspar** [comment]
Hi there, it turns out this stopped the error but doesn't really work effectively anymore.  The variable is used to avoid an additional unnecessary download.  Not sure how that would be fixed but 

`
$maven_version = undef
'

allowed the code to run and execute the download

**Comment by MikeCaspar** [comment]
Argh.. As it turns out, this works Once.
After that you receive an error that it cannot be re-assigned.

I'll do some searching to see where this variable is set (maybe in facter?).

Hopefully, this gets you through this for now.

**Comment by MikeCaspar** [comment]
OK.. Played around for a while.
I've found a fix that seems to work well for now and could be migrated to the module itself...

in site.pp...

`
if !getvar('::maven_version') {
  $maven_version = undef
}
`

The problem seems to arise from maven_version coming from facts.  If the fact exists, the variable is available.  If the fact does not yet exist, the variable won't exist. 

A likely better alternative might be to adjust maven.pp to check for the fact instead of the global variable (as per the above command) but I'm too new to puppet to know if this is a viable option.

**Comment by af6140** [comment]
Yes, the strict_variable parameter controls whether use uninitialized variable in puppet,  Googled around, there is no elegant solution. I had issue withe getvar also, guess different version of puppet may behave differently, it may also depends on whether you use future_parser.

**Comment by af6140** [comment]
Are we the only ones have the issues? Look at the puppet forge, this module is the most popular maven one, wondering what others experienced.

**Comment by MikeCaspar** [comment]
I have seen this type of thing with a few modules now.  

Although I am relatively new to puppet, from what I can see is that the problem is related to "strict variables" as a new approach and defining the scope of a variable properly when accessed. This seems to become more important when referencing facts.

I often see errors as well as it relates to $schedule when trying new modules.  I hope that helps.

I have learned to start being careful about using ::  when needed to reference higher level variables when needed. Could it simply be related to variable scope?

Sorry I can't help more.. maybe this weekend I might do a fork and play around a bit  to see if I can figure it out.

**Comment by cfiehe** [comment]
I have created a pull request that should ensure module's compliance with strict_variables=true. I had same trouble with Puppet 3.5.0 and Puppet 3.6.0, but now the change passes all tests.
